README.org
#+title: Documentation de l'algorithme d'orientation Covid-19

=Version: 2020-06-09=

** Introduction et contexte

Ce dépôt contient une présentation de l'algorithme d'orientation
Covid-19 officiellement publié par le ministère des Solidarités et de
la Santé.

L'algorithme est défini par un comité scientifique, appelé CovidTélé,
réunissant des médecins et co-piloté par l'AP-HP et l'Institut
Pasteur.

Afin que toutes les initiatives qui participent à la gestion de la
crise sanitaire s'appuient sur les mêmes recommendations, le ministère
des Solidarités et de la Santé conseille la réutilisation de cet
algorithme.

Voir le [[https://delegation-numerique-en-sante.github.io/covid19-algorithme-orientation/][site d'information générale]] produit à partir de ce dépôt.

[[file:doc-algo-orientation-covid19.png]]

** Présentation générale de la documentation

Dans ce dépôt, il faut bien distinguer plusieurs éléments :

- le *questionnaire*, composé des questions possibles et des réponses possibles pour chaque question ;
- le *parcours*, qui est une version /ordonnée/ du questionnaire
- les *données saisies* par l'utilisateurs (par exemple l'âge) ;
- les *données calculées* à partir des données saisies (par exemple la tranche d'âge) ;
- l'*algorithme*, qui est le calcul sur les /données saisies et calculées/ déterminant le message d'orientation.

Le /questionnaire/ propose un /parcours/ utilisateur qui permet de
recevoir des informations (les /données saisies/), d'en déduire des
valeurs intermédiaires (les /données calculées/) et de proposer un
message d'orientation, calculé à la fois à partir des valeurs saisies
et des valeurs intermédiaires.

Ces éléments donnent lieu à plusieurs présentations:

- Le [[https://delegation-numerique-en-sante.github.io/covid19-algorithme-orientation/algorithme-orientation-covid19.html][document officiel]] présente le questionnaire et l'algorithme de façon linéaire et synthétique.
- La page d'[[file:implementation.org][implémentation]] donne des détails sur les données saisies, les données calculées et leur sauvegarde pour partage.
- La page de [[file:pseudo-code.org][pseudo-code]] présente le questionnaire, le parcours et l'algorithme de façon analytique.

** Document officiel

Le document officiel de l'algorithme d'orientation Covid-19 est publié par le ministère des Solidarités et de la Santé.  Il est [[https://delegation-numerique-en-sante.github.io/covid19-algorithme-orientation/algorithme-orientation-covid19.html][lisible en ligne]] et [[https://esante.gouv.fr/algorithme-orientation][téléchargeable ici]].

** Référencement

Toutes les applications d'information, d'orientation et de triage qui
réutilisent cet algorithme sont référencées sur le [[https://solidarites-sante.gouv.fr/soins-et-maladies/maladies/maladies-infectieuses/coronavirus/coronavirus-questions-reponses][site du ministère
des Solidarités et de la Santé]].

Si vous implémentez cet algorithme et souhaitez partager vos données,
renseignez [[http://www.sesam-vitale.fr/web/sesam-vitale/recensement-innovations-covid-19][ce formulaire en ligne]] ou écrivez-nous à l'adresse
=mobilisation-covid@sante.gouv.fr= pour que votre initiative soit
référencée sur le site du ministère des Solidarité et de la Santé.

Toute implémentation référencée doit impérativement :

1. respecter [[file:pseudo-code.org][l'algorithme d'orientation]] Covid-19 ;
2. utiliser le [[https://github.com/Delegation-numerique-en-sante/covid19-algorithme-orientation-check/blob/master/schema.json][schéma de données]] officiel ;
3. et suivre les recommandations d'[[file:implementation.org][implémentation]].

Pour savoir si les données csv que vous produisez sont correctes, vous
pouvez utiliser cet [[https://github.com/Delegation-numerique-en-sante/covid19-algorithme-orientation-check][outil de validation]].

** Ce que contient ce dépôt

- un [[file:guide-evaluation-implementations.md][guide pour aider à l'évaluation des implémentations]] ;
- une présentation des [[file:pseudo-code.org][éléments formels de l'algorithme]] à respecter ;
- des recommandations pour le [[file:referencement.org][référencement]] de votre application ;
- des recommandations pour l'[[file:implementation.org][implémentation]] de l'algorithme ;
- des [[file:exemples.org][exemples]] d'implémentation et de mise en oeuvre d'une API ;
- une [[file:preuve-formelle/][validation formelle]] de l'algorithme.

Et aussi :

- le code source du [[https://delegation-numerique-en-sante.github.io/covid19-algorithme-orientation/demonstrateur.html][démonstrateur]] de l'algorithme ;
- les [[file:INSTALL.org][instructions]] à destination des développeurs de ce dépôt.

** Contribuer

Les contributions sont les bienvenues, sous forme d'/issues/ ou de /pull
requests/.

** Remerciements

Merci à Denis Merigoux pour la contribution de la [[file:preuve-formelle/][preuve formelle]].

Merci à Simon Chignard pour ses conseils sur la contextualisation des
informations concernant cette documentation.

Merci à Jean-Matthieu Schertzer d'être intervenu pour signaler un
problème.  Vous pouvez utilement consulter l'[[https://datajms.com/post/covid_scoring_system_presentation/][article de blog]] qu'il
consacre à une version simplifiée de l'algorithme de décision et à une
comparaison entre l'approche par arbre de décision et l'approche par
système de points, toutes deux logiquement équivalentes.

** Licence

2020 Délégation ministérielle du numérique en santé et les contributeurs du dépôt.

Le code source de ce dépôt est publié sous licence [[file:LICENSE][EPL-2.0]].

Le contenu du [[file:website/algorithme-orientation-covid19.org][document de présentation]] de l'algorithme est publié sous [[file:LICENSE.Etalab-2.0.md][licence Ouverte 2.0]].


ChangeLog.org
* 2020-06-09

** Correction de l'algorithme

Une [[https://github.com/Delegation-numerique-en-sante/covid19-algorithme-orientation/pull/49][correction]] a été apportée à l'une des branches de l'algorithme.

** Mise à jour des questions

Plusieurs questions ont été reformulées :

- /Avez-vous une toux ou votre toux habituelle s’est-elle modifiée ces derniers jours ?/
- /Avez-vous noté une perte ou une forte diminution de votre goût ou de votre odorat ces derniers jours ?/
- /Avez-vous un mal de gorge ou des douleurs musculaires ou des courbatures inhabituelles ou des maux de tête inhabituels ces derniers jours ?/
  - Si la réponse est OUI, afficher /Si vous avez des maux de tête qui ne disparaissent pas après la prise de médicaments anti-douleurs (dont paracétamol) et que vous décririez comme intolérables : Appelez le 15./
- /Êtes-vous une femme enceinte au 3ème trimestre de votre grossesse  ?/
- /Avez-vous des antécédents de maladie cardiovasculaire : hypertension artérielle compliquée (avec complications cardiaques, rénales), accident vasculaire cérébral, maladie coronaire (infarctus), chirurgie cardiaque, insuffisance cardiaque avec essoufflement au moindre effort ?/
- /Avez-vous une cirrhose ?/  [À la place de la question sur la maladie du foi.]
- /Avez-vous un diabète mal équilibré ou avec des complications (yeux, reins) ?/
- /Avez-vous une insuffisance rénale chronique avec besoin de faire de la dialyse ?/
- /Avez-vous un cancer évolutif sous traitement (hors hormonothérapie) ?/
- /Avez-vous une maladie respiratoire chronique (bronchopneumopathie obstructive, asthme sévère, fibrose pulmonaire, syndrome d’apnées du sommeil, mucoviscidose) ou êtes-vous suivi par un pneumologue ?/
- /Avez-vous une immunodépression, par exemple : médicamenteuse (chimiothérapie anti cancéreuse, traitement immunosuppresseur, biothérapie et/ou corticothérapie à dose immunosuppressive depuis plus de 15 jours) ; infection à VIH non contrôlée ou avec des CD4 <200/mm3 ; consécutive à une greffe d'organe solide ou de cellules souches hématopoïétiques ; liée à une hémopathie maligne en cours de traitement ?/

Une question a été ajoutée :

- /Avez-vous une drépanocytose homozygote (forme majeure) ou j’ai bénéficié d’une splénectomie (ablation de la rate) à cause de la drépanocytose ?/

Une question a été supprimée :

- /Prenez-vous un traitement immunosuppresseur ?/

** Mise à jour des variables exigées

- La variable =sickle_cell= correspondant à la question sur la
  drépanocytose a été ajoutée dans la liste des données exigées.

- Les variables =immunosuppressant_drug= et =immunosuppressant_drug_algo=
  ne sont plus exigées, car la question /Prenez-vous un traitement
  immunosuppresseur ?/ a été supprimée.

** Harmonisation en « Covid-19 »

L'appellation « Covid-19 » est désormais utilisée partout dans la documentation.

* 2020-05-10

** La limite pour la tranche d'âge supérieur est à 65 ans au lieu de 70 ans

Les valeurs possibles pour la variable =age_range= sont désormais
=inf_15=, =from_15_to_49=, =from_50_to_64= et =sup_65=.

Voir le [[https://github.com/Delegation-numerique-en-sante/covid19-algorithme-orientation-check/blob/master/schema.json][schéma de données]] mis à jour.

** Tous les messages d'orientation FIN2 ont été remplacées par ceux de la FIN6

Le message d'orientation =FIN2= était :

: Votre situation peut relever d’un COVID 19 qu’il faut surveiller. Si
: de nouveaux symptômes apparaissent, refaites le test ou consultez
: votre médecin. Nous vous conseillons de rester à votre domicile.

Tous ces messages sont désormais remplacés par celui de la =FIN6= :

: Votre situation peut relever d’un COVID 19. Demandez une
: téléconsultation ou un médecin généraliste ou une visite à domicile
: (SOS médecins, etc.) En attendant et pour casser les chaînes de
: transmission, nous vous conseillons de vous isoler et de respecter les
: gestes barrières pour protéger vos proches.

L'orientation =FIN2= disparaît donc de l'algorithme d'orientation.

** Les messages d'orientation pour FIN3, FIN4, FIN6, FIN7 ont été modifiés

Pour tous ces messages, la phrase suivante à été ajoutée à la fin :

: En attendant et pour casser les chaînes de transmission, nous vous
: conseillons de vous isoler et de respecter les gestes barrières pour
: protéger vos proches.

* 2020-04-29

** Ajout d'un identifiant unique

Le tableau des données du fichier =implementation.org= contient une
nouvelle ligne pour l'ajout d'un identifiant unique, optionnel.

Vous reporter au [[https://github.com/Delegation-numerique-en-sante/covid19-algorithme-orientation-check/][dépôt du schéma de données]] pour les détails.

** Nouvelles valeurs acceptables pour la variable "orientation"

 - less_15
 - home_surveillance
 - consultation_surveillance_1
 - consultation_surveillance_2
 - SAMU
 - consultation_surveillance_3
 - consultation_surveillance_4
 - surveillance

* 2020-04-17

** L'arbre de décision a été corrigé pour les 4ème et 5ème clauses

La 4ème clause disait:

: "IF fièvre OR (diarrhée OR (toux AND douleurs) OR (toux AND anosmie))"

Elle dit désormais:

: IF fièvre OR (diarrhée OR (toux AND douleurs) OR (toux AND anosmie) OR (douleurs AND anosmie))

La 5ème clause disait:

: IF toux OR douleurs OR anosmie
 
Elle dit désormais:

: IF toux XOR douleurs XOR anosmie

Voir le [[file:pseudo-code.org][pseudo-code]].

** Une seule question pour la fièvre

La fièvre était demandée en deux temps : d'abord une question pour
savoir si le répondant avait de la fièvre, une autre pour demander la
température.  Il n'y a désormais qu'une seule question, celle sur la
température.

En conséquence, les solutions ne doivent plus envoyer de valeur ~fever~
mais seulement ~fever_algo~ et ~temperature_cat~ : voir [[file:implementation.org][l'implémentation]].

À noter que si le répondant indique "Je ne sais pas", cela est
considéré ensuite dans l'algorithme comme indiquant de la fièvre.

** Clarification sur la question du code postal


src/clj/choices/repl/clojure/preprocess.clj
;; Fonction de calcul de l'indice de masse corporelle
(defn compute-bmi [weight height]
  (/ weight (Math/pow (/ height 100.0) 2)))

;; Fonction pour réduire l'âge à la tranche d'âge
(defn get-age-range [reponse]
  (cond (< (:age reponse) 15) "inf_15"
        (< (:age reponse) 50) "from_15_to_49"
        (< (:age reponse) 70) "from_50_to_64"
        :else                 "sup_65"))

;; Fonction pour mettre le bmi, age-range dans la réponse, et
;; incrémenter les facteurs pronostiques ou de gravité
(defn preprocess-scores [reponse]
  (let [bmi-val   (compute-bmi (:weight reponse)
                               (:height reponse))
        age-range (get-age-range reponse)
        reponse   (merge reponse {:bmi bmi-val})
        reponse   (merge reponse {:age-range age-range})
        reponse   (update-in reponse [:pronostic-factors]
                             #(if (= (:age-range reponse) "sup_65") (inc %) %))
        reponse   (update-in reponse [:pronostic-factors]
                             #(if (>= bmi-val 30) (inc %) %))
        reponse   (dissoc reponse :weight :height :age)]
    ;; Renvoie la réponse remise en forme:
    reponse))
